### PR TITLE
[PDI-17801] Different output fields of Table Input for MySQL and MariaDB driver

### PR DESCRIPTION
**Pull request by jpruiz114**
@pentaho/tatooine @pentaho-lmartins @gryphendowre @wseyler

[PDI-17801] Moving MySQL logic from db class to db dialects
[PDI-17801] Adding unit tests for MariaDB legacy column feature
[PDI-17801] Adding unit tests for MySQL legacy column feature
[PDI-17801] Refactoring MariaDB and MySQL unit tests for legacy column name methods
[PDI-17801] Changing try catch block to match code style
[PDI-17801] Changing try catch code style

### CODE REVIEW

File: core/src/test/java/org/pentaho/di/core/database/MySQLDatabaseMetaTest.java [hunk 1]
```diff
@@ -38,16 +43,59 @@
...
+  private DatabaseMetaData databaseMetaData;
+
+  private ResultSetMetaData resultSetMetaData, resultSetMetaDataException;
+
+  @Rule
+  public ExpectedException expectedException = ExpectedException.none();
+
   @Before
-  public void setupBefore() {
+  public void setupBefore() throws Exception {
```
Comment (pentaho-lmartins) [inline core/src/test/java/org/pentaho/di/core/database/MySQLDatabaseMetaTest.java]: Please change this method name to just setup. Like in the MariaDB test suite.

File: core/src/test/java/org/pentaho/di/core/database/MySQLDatabaseMetaTest.java [hunk 2]
```diff
@@ -38,16 +43,59 @@
...
     nativeMeta.setAccessType( DatabaseMeta.TYPE_ACCESS_NATIVE );
+
     odbcMeta = new MySQLDatabaseMeta();
     odbcMeta.setAccessType( DatabaseMeta.TYPE_ACCESS_ODBC );
 
+    databaseMetaData = mock( DatabaseMetaData.class );
+
+    resultSetMetaData = mock( ResultSetMetaData.class );
+
+    given( resultSetMetaData.getColumnLabel( 1 ) ).willReturn( "NUMBER" );
```
Comment (pentaho-lmartins) [inline core/src/test/java/org/pentaho/di/core/database/MySQLDatabaseMetaTest.java]: Same comments as above.

**Comment by pentaho-lmartins** [comment]
And there are some failed unit tests... You need to look into those.

File: core/src/test/java/org/pentaho/di/core/database/MySQLDatabaseMetaTest.java (hunk 1 above)
Comment (jpruiz114) [reply core/src/test/java/org/pentaho/di/core/database/MySQLDatabaseMetaTest.java]: Hey @pentaho-lmartins: This method has been named like that since before I added the extra methods to the MySQL unit test class. There's a sub-class depending on that method as well, InfiniDbDatabaseMetaTest. I would feel more comfortable leaving it as it is. Let me know your thoughts.

File: core/src/test/java/org/pentaho/di/core/database/MariaDBDatabaseMetaTest.java [hunk 3]
```diff
@@ -21,24 +21,92 @@
...
+    expectedException.expect( Exception.class );
+
+    nativeMeta.getLegacyColumnName( null, null, 1 );
+  }
+
+  @Test
+  public void testGetLegacyColumnNameDatabaseException() throws Exception {
+    MariaDBDatabaseMeta nativeMeta = new MariaDBDatabaseMeta();
+
+    expectedException.expect( KettleDatabaseException.class );
```
[the remark this answers is not in the document]
Comment (jpruiz114) [reply core/src/test/java/org/pentaho/di/core/database/MariaDBDatabaseMetaTest.java]: Changed it to `@Test( expected = KettleDatabaseException.class )`

File: core/src/test/java/org/pentaho/di/core/database/MariaDBDatabaseMetaTest.java [hunk 4]
```diff
@@ -21,24 +21,92 @@
...
+
+    }
+  }
 
   @Test
-  public void testMysqlOverrides() throws Exception {
+  public void testGetLegacyColumnNameNullParametersException() throws Exception {
+    MariaDBDatabaseMeta nativeMeta = new MariaDBDatabaseMeta();
+
+    expectedException.expect( Exception.class );
```
[the remark this answers is not in the document]
Comment (jpruiz114) [reply core/src/test/java/org/pentaho/di/core/database/MariaDBDatabaseMetaTest.java]: Changed it to `@Test( expected = KettleDatabaseException.class )`

File: core/src/main/java/org/pentaho/di/core/database/MariaDBDatabaseMeta.java [hunk 5]
```diff
@@ -62,4 +63,26 @@
...
+   * @param rsMetaData
+   * @param index
+   * @return The column label.
+   * @throws KettleDatabaseException
+   */
+  @Override public String getLegacyColumnName( DatabaseMetaData dbMetaData, ResultSetMetaData rsMetaData, int index ) throws KettleDatabaseException {
+    String columnName;
+
+    try {
+      columnName = rsMetaData.getColumnLabel( index );
```
[the remark this answers is not in the document]
Comment (jpruiz114) [reply core/src/main/java/org/pentaho/di/core/database/MariaDBDatabaseMeta.java]: Changed to return the column name inside the try block.

File: core/src/test/java/org/pentaho/di/core/database/MariaDBDatabaseMetaTest.java [hunk 6]
```diff
@@ -21,24 +21,92 @@
...
+    MariaDBDatabaseMeta nativeMeta = new MariaDBDatabaseMeta();
+
+    try {
+      assertEquals( "NUMBER", nativeMeta.getLegacyColumnName( databaseMetaData, resultSetMetaData, 1 ) );
+      assertEquals( "NAME", nativeMeta.getLegacyColumnName( databaseMetaData, resultSetMetaData, 2 ) );
+      assertEquals( "LAST_NAME", nativeMeta.getLegacyColumnName( databaseMetaData, resultSetMetaData, 3 ) );
+      assertEquals( "FIRST_NAME", nativeMeta.getLegacyColumnName( databaseMetaData, resultSetMetaData, 4 ) );
+      assertEquals( "DB", nativeMeta.getLegacyColumnName( databaseMetaData, resultSetMetaData, 5 ) );
+      assertEquals( "NoAliasText", nativeMeta.getLegacyColumnName( databaseMetaData, resultSetMetaData, 6 ) );
+    } catch ( KettleDatabaseException kettleDatabaseException ) {
```
[the remark this answers is not in the document]
Comment (jpruiz114) [reply core/src/test/java/org/pentaho/di/core/database/MariaDBDatabaseMetaTest.java]: Removed try-catch blocks so we can see the tests fail.

File: core/src/main/java/org/pentaho/di/core/database/MariaDBDatabaseMeta.java [hunk 7]
```diff
@@ -62,4 +63,26 @@
...
+    } catch ( SQLException sqlException ) {
+      throw new KettleDatabaseException( "Something went wrong trying to get the legacy column name", sqlException );
```
[the remark this answers is not in the document]
Comment (jpruiz114) [reply core/src/main/java/org/pentaho/di/core/database/MariaDBDatabaseMeta.java]: Texts have been moved to the properties file.

File: core/src/main/java/org/pentaho/di/core/database/MySQLDatabaseMeta.java [hunk 8]
```diff
@@ -480,4 +484,27 @@ public int getMaxVARCHARLength() {
...
+   * Returns the column name for a MySQL field checking if the driver major version is "greater than" or "lower or equal" to 3.
+   *
+   * @param dbMetaData
+   * @param rsMetaData
+   * @param index
+   * @return The column label if version is greater than 3 or the column name if version is lower or equal to 3.
+   * @throws KettleDatabaseException
+   */
+  public String getLegacyColumnName( DatabaseMetaData dbMetaData, ResultSetMetaData rsMetaData, int index ) throws KettleDatabaseException {
+    String columnName;
```
[the remark this answers is not in the document]
Comment (jpruiz114) [reply core/src/main/java/org/pentaho/di/core/database/MySQLDatabaseMeta.java]: Changed to return the column name inside the try block.

File: core/src/test/java/org/pentaho/di/core/database/MariaDBDatabaseMetaTest.java [hunk 9]
```diff
@@ -21,24 +21,92 @@
...
+  @Rule
+  public ExpectedException expectedException = ExpectedException.none();
+
+  @Before
+  public void setUp() throws Exception {
+    databaseMetaData = mock( DatabaseMetaData.class );
+
+    resultSetMetaData = mock( ResultSetMetaData.class );
+
+    given( resultSetMetaData.getColumnLabel( 1 ) ).willReturn( "NUMBER" );
```
[the remark this answers is not in the document]
Comment (jpruiz114) [reply core/src/test/java/org/pentaho/di/core/database/MariaDBDatabaseMetaTest.java]: Syntax changed from "given-willReturn" to "doReturn-when".

File: core/src/test/java/org/pentaho/di/core/database/MariaDBDatabaseMetaTest.java [hunk 10]
```diff
@@ -21,24 +21,92 @@
...
+    given( resultSetMetaData.getColumnLabel( 2 ) ).willReturn( "NAME" );
+    given( resultSetMetaData.getColumnLabel( 3 ) ).willReturn( "LAST_NAME" );
+    given( resultSetMetaData.getColumnLabel( 4 ) ).willReturn( "FIRST_NAME" );
+    given( resultSetMetaData.getColumnLabel( 5 ) ).willReturn( "DB" );
+    given( resultSetMetaData.getColumnLabel( 6 ) ).willReturn( "NoAliasText" );
+
+    resultSetMetaDataException = mock( ResultSetMetaData.class );
+
+    when( resultSetMetaDataException.getColumnLabel( 1 ) ).thenThrow( new SQLException() );
```
[the remark this answers is not in the document]
Comment (jpruiz114) [reply core/src/test/java/org/pentaho/di/core/database/MariaDBDatabaseMetaTest.java]: Syntax changed from "when-thenThrow" to "doThrow-when".

File: core/src/test/java/org/pentaho/di/core/database/MySQLDatabaseMetaTest.java (hunk 2 above)
Comment (jpruiz114) [reply core/src/test/java/org/pentaho/di/core/database/MySQLDatabaseMetaTest.java]: Syntax changed from "given-willReturn" to "doReturn-when".

Comment (jpruiz114) [reply core/src/test/java/org/pentaho/di/core/database/MySQLDatabaseMetaTest.java]: Syntax changed from "when-thenThrow" to "doThrow-when".

File: core/src/main/java/org/pentaho/di/core/database/Database.java [hunk 11]
```diff
@@ -721,11 +721,8 @@ public synchronized void closeConnectionOnly() throws KettleDatabaseException {
    * @throws KettleDatabaseException
    */
   public void cancelQuery() throws KettleDatabaseException {
-    // Canceling statements only if we're not streaming results on MySQL with
-    // the v3 driver
-    //
-    if ( databaseMeta.isMySQLVariant()
-      && databaseMeta.isStreamingResults() && getDatabaseMetaData().getDriverMajorVersion() == 3 ) {
+    // Canceling statements only if we're not streaming results on MySQL with the v3 driver
```
[the remark this answers is not in the document]
Comment (jpruiz114) [reply core/src/main/java/org/pentaho/di/core/database/Database.java]: Agreed.

File: core/src/test/java/org/pentaho/di/core/database/MariaDBDatabaseMetaTest.java [hunk 12]
```diff
@@ -21,24 +21,92 @@
...
+    when( resultSetMetaDataException.getColumnLabel( 5 ) ).thenThrow( new SQLException() );
+    when( resultSetMetaDataException.getColumnLabel( 6 ) ).thenThrow( new SQLException() );
+  }
+
+  @Test
+  public void testGetLegacyColumnName() {
+    MariaDBDatabaseMeta nativeMeta = new MariaDBDatabaseMeta();
+
+    try {
+      assertEquals( "NUMBER", nativeMeta.getLegacyColumnName( databaseMetaData, resultSetMetaData, 1 ) );
```
[the remark this answers is not in the document]
Comment (jpruiz114) [reply core/src/test/java/org/pentaho/di/core/database/MariaDBDatabaseMetaTest.java]: Wrote individual tests for each assert of MariaDB unit test class.

File: core/src/test/java/org/pentaho/di/core/database/MariaDBDatabaseMetaTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (jpruiz114) [reply core/src/test/java/org/pentaho/di/core/database/MariaDBDatabaseMetaTest.java]: Removed testing of fields from index 2 to index 6.

File: core/src/main/java/org/pentaho/di/core/database/Database.java [hunk 13]
```diff
@@ -2646,26 +2643,24 @@ private ValueMetaInterface getValueFromSQLType( ResultSetMetaData rm, int i, boo
...
 
     // Extract the name from the result set meta data...
-    //
-    String name;
-    if ( databaseMeta.isMySQLVariant() && getDatabaseMetaData().getDriverMajorVersion() > 3 ) {
-      name = new String( rm.getColumnLabel( i ) );
-    } else {
-      name = new String( rm.getColumnName( i ) );
+
+    String name = null;
```
[the remark this answers is not in the document]
Comment (jpruiz114) [reply core/src/main/java/org/pentaho/di/core/database/Database.java]: Changed as it was before.

File: core/src/test/java/org/pentaho/di/core/database/MySQLDatabaseMetaTest.java (hunk 2 above)
Comment (jpruiz114) [reply core/src/test/java/org/pentaho/di/core/database/MySQLDatabaseMetaTest.java]: Created separate tests for each field.

**Comment by jpruiz114** [comment]
@pentaho/tatooine @pentaho-lmartins @gryphendowre @wseyler Completed suggested changes, fixed failing tests.

[PDI-17801] Moving MySQL logic from db class to db dialects
[PDI-17801] Adding unit tests for MariaDB legacy column feature
[PDI-17801] Adding unit tests for MySQL legacy column feature
[PDI-17801] Refactoring MariaDB and MySQL unit tests for legacy column name methods
[PDI-17801] Changing try catch block to match code style
[PDI-17801] Changing try catch code style
[PDI-17801] Refactoring code to match code review suggestions
[PDI-17801] Adding reinitialization of db meta obj for unit test edge cases
[PDI-17801] Extending details of exception message for mysql unit tests
[PDI-17801] Adding changes to MySQL unit tests for legacy column name feature
[PDI-17801] Adding changes to MariaDB and MySQL unit test classes
[PDI-17801] Adding changes to MySQL unit tests
[PDI-17801] Refactor MariaDB and MySQL unit tests, removing non used libraries and i18n texts
[PDI-17801] Refactoring Database class and adding back else case for non mysql variant
[PDI-17801] Removing non used library from mysql db meta class
[PDI-17801] Adding separate unit tests for each field on version 3 and 5 of mysql connector
[PDI-17801] Adding details to MariaDB exception catch block

**Comment by jpruiz114** [comment]
@pentaho/tatooine @pentaho-lmartins @gryphendowre @wseyler Completed suggested changes, fixed failing tests.

[PDI-17801] Moving MySQL logic from db class to db dialects
[PDI-17801] Adding unit tests for MariaDB legacy column feature
[PDI-17801] Adding unit tests for MySQL legacy column feature
[PDI-17801] Refactoring MariaDB and MySQL unit tests for legacy column name methods
[PDI-17801] Changing try catch block to match code style
[PDI-17801] Changing try catch code style
[PDI-17801] Refactoring code to match code review suggestions
[PDI-17801] Adding reinitialization of db meta obj for unit test edge cases
[PDI-17801] Extending details of exception message for mysql unit tests
[PDI-17801] Adding changes to MySQL unit tests for legacy column name feature
[PDI-17801] Adding changes to MariaDB and MySQL unit test classes
[PDI-17801] Adding changes to MySQL unit tests
[PDI-17801] Refactor MariaDB and MySQL unit tests, removing non used libraries and i18n texts
[PDI-17801] Refactoring Database class and adding back else case for non mysql variant
[PDI-17801] Removing non used library from mysql db meta class
[PDI-17801] Adding separate unit tests for each field on version 3 and 5 of mysql connector
[PDI-17801] Adding details to MariaDB exception catch block

**Comment by jpruiz114** [comment]
@pentaho/tatooine @pentaho-lmartins @gryphendowre @wseyler Completed suggested changes, fixed failing tests.

[PDI-17801] Moving MySQL logic from db class to db dialects
[PDI-17801] Adding unit tests for MariaDB legacy column feature
[PDI-17801] Adding unit tests for MySQL legacy column feature
[PDI-17801] Refactoring MariaDB and MySQL unit tests for legacy column name methods
[PDI-17801] Changing try catch block to match code style
[PDI-17801] Changing try catch code style
[PDI-17801] Refactoring code to match code review suggestions
[PDI-17801] Adding reinitialization of db meta obj for unit test edge cases
[PDI-17801] Extending details of exception message for mysql unit tests
[PDI-17801] Adding changes to MySQL unit tests for legacy column name feature
[PDI-17801] Adding changes to MariaDB and MySQL unit test classes
[PDI-17801] Adding changes to MySQL unit tests
[PDI-17801] Refactor MariaDB and MySQL unit tests, removing non used libraries and i18n texts
[PDI-17801] Refactoring Database class and adding back else case for non mysql variant
[PDI-17801] Removing non used library from mysql db meta class
[PDI-17801] Adding separate unit tests for each field on version 3 and 5 of mysql connector
[PDI-17801] Adding details to MariaDB exception catch block

**Comment by jpruiz114** [comment]
@pentaho/tatooine @pentaho-lmartins @gryphendowre @wseyler Completed suggested changes, fixed failing tests.

[PDI-17801] Moving MySQL logic from db class to db dialects
[PDI-17801] Adding unit tests for MariaDB legacy column feature
[PDI-17801] Adding unit tests for MySQL legacy column feature
[PDI-17801] Refactoring MariaDB and MySQL unit tests for legacy column name methods
[PDI-17801] Changing try catch block to match code style
[PDI-17801] Changing try catch code style
[PDI-17801] Refactoring code to match code review suggestions
[PDI-17801] Adding reinitialization of db meta obj for unit test edge cases
[PDI-17801] Extending details of exception message for mysql unit tests
[PDI-17801] Adding changes to MySQL unit tests for legacy column name feature
[PDI-17801] Adding changes to MariaDB and MySQL unit test classes
[PDI-17801] Adding changes to MySQL unit tests
[PDI-17801] Refactor MariaDB and MySQL unit tests, removing non used libraries and i18n texts
[PDI-17801] Refactoring Database class and adding back else case for non mysql variant
[PDI-17801] Removing non used library from mysql db meta class
[PDI-17801] Adding separate unit tests for each field on version 3 and 5 of mysql connector
[PDI-17801] Adding details to MariaDB exception catch block

**Comment by jpruiz114** [comment]
@pentaho-lmartins please hold the review of this ticket. I am doing some additional testing with the latest build of PDI client 8.3 and I'm getting some errors which I was not getting with a previous build.
